storage: Transfer lease to least-loaded store when rebalancing replicas

a-robinson opened the pull request:
This logic was totally busted before due to indexing into the wrong
replicas slice, meaning we would often transfer the lease to the wrong
store.

Release note: None

a-robinson committed:
storage: Transfer lease to least-loaded store when rebalancing replicas

This logic was totally busted before due to indexing into the wrong
replicas slice, meaning we would often transfer the lease to the wrong
store.

Also add a proper test for the replica/lease selection logic.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/30256)
<!-- Reviewable:end -->

BramGruneir commented:
LGTM.

Any way to add a test to make sure the logic is correct?

On Sat, Sep 15, 2018, 15:44 cockroach-teamcity <notifications@github.com>
wrote:

> This change is [image: Reviewable]
> <https://reviewable.io/reviews/cockroachdb/cockroach/30256>
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/30256#issuecomment-421623631>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABihuVf60jYH7RxX4mFv_ScigUoINyBdks5ubViAgaJpZM4WqgvQ>
> .
>

a-robinson committed:
storage: Test and slightly improve store rebalancer decisions

Further protect against picking rebalance targets that are somewhat full
by both extending the allocator_scorer convergeScore logic to add an
extra score for overfull stores, and add a couple fallback checks to
back up the convergeScore logic when needed.

Release note: None

a-robinson reviewed (commented):
Yeah, absolutely. Done (along with a couple tweaks to slightly improve decisions). PTAL

a-robinson commented:
ping @BramGruneir

BramGruneir reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 2 of 2 files at r1, 3 of 3 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/30256#-:-LMmuHtgChHcDlOvtBoX:bnfp4nl)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)



<!-- Sent from Reviewable.io -->

a-robinson commented:
bors r+

a-robinson commented:
Updated one line in the test to fix merge skew with https://github.com/cockroachdb/cockroach/pull/30143.

bors r+

a-robinson commented:
Heh, now that I've rebased on master it's failing due to https://github.com/cockroachdb/cockroach/issues/30435. I'll fix that up before continuing with this.

a-robinson commented:
I'll wait on this until #30422 is in, since it fixes the test failures.
